ledger/participant-integration-api/src/main/scala/platform/index/IndexServiceImpl.scala

```diff
@@ -467,6 +475,67 @@ private[index] class IndexServiceImpl(
 }
 
 object IndexServiceImpl {
+  private val logger = ContextualizedLogger.get(getClass)
+
+  private[index] def unknownTemplatesOrInterfaces(
+      domainTransactionFilter: domain.TransactionFilter,
+      metadata: PackageMetadata,
+  ) =
+    (for {
+      (_, inclusiveFilterOption) <- domainTransactionFilter.filtersByParty.iterator
+      inclusiveFilter <- inclusiveFilterOption.inclusive.iterator
+      unknownInterfaces =
+        inclusiveFilter.interfaceFilters
+          .collect {
+            case interfaceFilter if !metadata.interfaces.contains(interfaceFilter.interfaceId) =>
+              Right(interfaceFilter.interfaceId)
+          }
+      unknownTemplates = inclusiveFilter.templateIds
+        .collect {
+          case templateId if !metadata.templates.contains(templateId) => Left(templateId)
+        }
+      unknownTemplateOrInterface <- unknownInterfaces ++ unknownTemplates
+    } yield unknownTemplateOrInterface).toList
+
+  private[index] def withValidatedFilter[T](
+      domainTransactionFilter: domain.TransactionFilter,
+      metadata: PackageMetadata,
+  )(
+      source: => Source[T, NotUsed]
+  )(implicit loggingContext: LoggingContext): Source[T, NotUsed] = {
+    implicit val errorLogger: DamlContextualizedErrorLogger =
+      new DamlContextualizedErrorLogger(logger, loggingContext, None)
+
+    val templatesOrInterfaces = unknownTemplatesOrInterfaces(domainTransactionFilter, metadata)
+
+    if (templatesOrInterfaces.nonEmpty) {
+      Source.failed(
+        invalidArgument(
+          invalidTemplateOrInterfaceMessage(templatesOrInterfaces)
+        )
+      )
+    } else
+      source
```

Comment:
I would either use curlies for both branches or for none 
```suggestion
    } else {
      source
    } 
```